base/multidimensional.jl

```diff
@@ -351,37 +376,39 @@ module IteratorsMD
         iterfirst, iterfirst
     end
     @inline function iterate(iter::CartesianIndices, state)
-        valid, I = __inc(state.I, first(iter).I, last(iter).I)
+        valid, I = __inc(state.I, first(iter).I, step.(iter.indices), last(iter).I)
         valid || return nothing
         return CartesianIndex(I...), CartesianIndex(I...)
     end
 
     # increment & carry
-    @inline function inc(state, start, stop)
-        _, I = __inc(state, start, stop)
+    # TODO: optimize this; it adds up about 5ns overhead
+    @inline function inc(state, start, step, stop)
```

Comment:
A better strategy to remove the overhead that I have in mind is to dispatch `inc`/`dec` on `StepRange` and `AbstractUnitRange`.

Reply:
Surprisingly, a0e64c3 also reduces the overhead that #35074 tries to solve.

```julia
function sumcart_iter(A)
    s = zero(eltype(A))
    for I in CartesianIndices(A)
        @inbounds s += A[I]
    end
    s
end
function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in ax2, i in ax1
        @inbounds s += A[i, j]
    end
    s
end
A = rand(4, 4)
using Test
@test sumcart_iter(A) == sumcart_manual(A)
using BenchmarkTools
@btime sumcart_iter($A)
@btime sumcart_manual($A)
```

```julia
julia> @btime sumcart_iter($A)
  7.316 ns (0 allocations: 0 bytes) # this PR
  14.839 ns (0 allocations: 0 bytes) # master
8.549672595595762
julia> @btime sumcart_manual($A)
  8.107 ns (0 allocations: 0 bytes)
8.549672595595762
```

and for the step range version:

```julia
function sumcart_iter(A)
    s = zero(eltype(A))
    R = CartesianIndices(A)
    Δ = CartesianIndex(ntuple(_->2, ndims(R)))
    for I in first(R):Δ:last(R)
        @inbounds s += A[I]
    end
    s
end

function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in first(ax2):2:last(ax2), i in first(ax1):2:last(ax1)
        @inbounds s += A[i, j]
    end
    s
end

A = rand(8, 8)
using Test
@test sumcart_iter(A) == sumcart_manual(A)
using BenchmarkTools
@btime sumcart_iter($A)
@btime sumcart_manual($A)
```

```julia
julia> @btime sumcart_iter($A)
  15.162 ns (0 allocations: 0 bytes)
9.210002105899234

julia> @btime sumcart_manual($A)
  27.261 ns (0 allocations: 0 bytes)
9.210002105899234
```